### PR TITLE
Fix checking required fields and signature

### PR DESCRIPTION
**Pull request by Hawk73**

### CODE REVIEW

File: lib/omniauth/strategies/telegram.rb [hunk 1]
```diff
@@ -81,14 +80,15 @@ def check_errors
...
-        FIELDS.all? { |f| request.params.include?(f) }
+        REQUIRED_FIELDS.all? { |f| request.params.include?(f) }
       end
-      
+
       def check_signature
         secret = OpenSSL::Digest::SHA256.digest(options[:bot_secret])
-        signature = HASH_FIELDS.map { |f| "%s=%s" % [f, request.params[f]] }.join("\n")
+        keys = (request.params.keys - ['hash']).sort
+        signature = keys.map { |key| "%s=%s" % [key, request.params[key]] }.join("\n")
```
Comment (Hawk73) [inline lib/omniauth/strategies/telegram.rb:89]: > Data-check-string is a concatenation of all received fields, sorted in alphabetical order

https://core.telegram.org/widgets/login

File: lib/omniauth/strategies/telegram.rb [hunk 2]
```diff
@@ -13,10 +13,9 @@ class Telegram
       option :bot_name, nil
       option :bot_secret, nil
       option :button_config, {}
-      
-      FIELDS      = %w[id first_name last_name username photo_url auth_date hash]
-      HASH_FIELDS = %w[auth_date first_name id last_name photo_url username]
-      
+
+      REQUIRED_FIELDS = %w[id username auth_date hash]
```
Comment (Hawk73) [inline lib/omniauth/strategies/telegram.rb:17]: Currently I don't get the `photo_url` param for my telegram account. So I simplified checking params.

File: lib/omniauth/strategies/telegram.rb [hunk 3]
```diff
@@ -81,14 +81,15 @@ def check_errors
...
       def check_fields
-        FIELDS.all? { |f| request.params.include?(f) }
+        REQUIRED_FIELDS.all? { |f| request.params.include?(f) }
       end
-      
+
       def check_signature
         secret = OpenSSL::Digest::SHA256.digest(options[:bot_secret])
-        signature = HASH_FIELDS.map { |f| "%s=%s" % [f, request.params[f]] }.join("\n")
+        keys = (request.params.keys & SIGNATURE_FIELDS).sort
```
Comment (Hawk73) [inline lib/omniauth/strategies/telegram.rb:89]: > Data-check-string is a concatenation of all received fields, sorted in alphabetical order

https://core.telegram.org/widgets/login

Also my app has another params too (locale etc) and that is why I added filter by `SIGNATURE_FIELDS`

File: lib/omniauth/strategies/telegram.rb [hunk 4]
```diff
@@ -13,10 +13,10 @@ class Telegram
       option :bot_name, nil
       option :bot_secret, nil
       option :button_config, {}
-      
-      FIELDS      = %w[id first_name last_name username photo_url auth_date hash]
-      HASH_FIELDS = %w[auth_date first_name id last_name photo_url username]
-      
+
+      REQUIRED_FIELDS = %w[id hash]
```
Comment (Hawk73) [inline lib/omniauth/strategies/telegram.rb:17]: Currently I don't get the `photo_url` param for my telegram account. So I simplified checking params.

File: lib/omniauth/strategies/telegram.rb (hunk 3 above)
Comment (phoet) [reply lib/omniauth/strategies/telegram.rb:89]: @Hawk73 is the documentation outdated? or where do the `locale` param etc come from?

Comment (Hawk73) [reply lib/omniauth/strategies/telegram.rb:89]: Please let me explain it. I add `locale` param to callback URL for saving current locale. So this param returns with telegrams params. That is why I added whitelist of params to check signature correctly.

Comment (phoet) [reply lib/omniauth/strategies/telegram.rb:89]: wouldn't that allow an attacker to pass arbitrary data via the callback url-params? i'm not super into the omniauth login widget, so please point it out if i'm thinking in the wrong direction here.

**Comment by phoet** [comment]
could you checkout 0.2.1 and see if that fixes your problems as well?

**Pull request closed without merging**